File: cookbook/route-handler.ts
/**
 * route-handler.ts — CopilotForge Cookbook Recipe
 *
 * WHAT THIS DOES:
 *   Demonstrates an Express route handler with Zod input validation,
 *   typed request/response, centralized error middleware, and proper
 *   HTTP status codes. Shows patterns for GET (list/detail), POST (create),
 *   and PATCH (update) endpoints.
 *
 * WHEN TO USE THIS:
 *   When building Express APIs and you want consistent request validation,
 *   error responses, and typed handlers — without a heavyweight framework.
 *
 * HOW TO RUN:
 *   1. npm install express zod
 *   2. npm install -D @types/express
 *   3. npx ts-node cookbook/route-handler.ts
 *
 * PREREQUISITES:
 *   - Node.js 18+
 *   - TypeScript 5+
 *   - express ^4.18
 *   - zod ^3.22
 *
 * PLATFORM NOTES:
 *   - Windows: Use backslashes in paths or path.join() (both shown in code)
 *   - macOS/Linux: Forward slashes work natively
 *   - Environment variables: Use $env:VAR (PowerShell) or export VAR (bash)
 */

import express, { Request, Response, NextFunction, Router } from "express";
import { z, ZodError, ZodSchema } from "zod";

// --- Validation Middleware ---

/**
 * Generic validation middleware that validates request body, query, or params
 * against a Zod schema. Returns 400 with structured errors on failure.
 *
 * Usage:
 *   router.post("/items", validate(createItemSchema, "body"), createItem);
 *   router.get("/items", validate(listQuerySchema, "query"), listItems);
 */
function validate(schema: ZodSchema, source: "body" | "query" | "params" = "body") {
  return (req: Request, res: Response, next: NextFunction): void => {
    const result = schema.safeParse(req[source]);

    if (!result.success) {
      const errors = result.error.errors.map((e) => ({
        field: e.path.join("."),
        message: e.message,
      }));

      res.status(400).json({
        error: {
          code: "VALIDATION_ERROR",
          message: "Request validation failed",
          details: errors,
        },
      });
      return;
    }

    // Replace with parsed (and coerced) values.
    req[source] = result.data;
    next();
  };
}

// --- Error Middleware ---

/** App-level error with status code and error code. */
class RouteError extends Error {
  readonly statusCode: number;
  readonly code: string;

  constructor(message: string, statusCode: number, code: string) {
    super(message);
    this.name = "RouteError";
    this.statusCode = statusCode;
    this.code = code;
  }
}

/**
 * Centralized error handler. Place at the end of your middleware chain.
 * Catches RouteError (operational) and unknown errors (500).
 */
function errorHandler(err: Error, _req: Request, res: Response, _next: NextFunction): void {
  if (err instanceof RouteError) {
    res.status(err.statusCode).json({
      error: { code: err.code, message: err.message },
    });
    return;
  }

  if (err instanceof ZodError) {
    const errors = err.errors.map((e) => ({
      field: e.path.join("."),
      message: e.message,
    }));
    res.status(400).json({
      error: { code: "VALIDATION_ERROR", message: "Validation failed", details: errors },
    });
    return;
  }

  // Unknown error — log and return generic 500.
  console.error("[ERROR]", err);
  res.status(500).json({
    error: { code: "INTERNAL_ERROR", message: "An unexpected error occurred" },
  });
}

// --- Schemas ---

const createItemSchema = z.object({
  name: z.string().min(1, "Name is required").max(100, "Name must be 100 chars or less"),
  description: z.string().max(500).optional(),
  price: z.number().positive("Price must be positive"),
  tags: z.array(z.string()).max(10).default([]),
});

const updateItemSchema = z.object({
  name: z.string().min(1).max(100).optional(),
  description: z.string().max(500).optional(),
  price: z.number().positive().optional(),
  tags: z.array(z.string()).max(10).optional(),
});

const listQuerySchema = z.object({
  page: z.coerce.number().int().min(1).default(1),
  pageSize: z.coerce.number().int().min(1).max(100).default(20),
  search: z.string().optional(),
});

const itemIdSchema = z.object({
  id: z.string().uuid("Invalid item ID format"),
});

// --- Types (inferred from schemas) ---

type CreateItemInput = z.infer<typeof createItemSchema>;
type UpdateItemInput = z.infer<typeof updateItemSchema>;
type ListQuery = z.infer<typeof listQuerySchema>;

interface Item {
  id: string;
  name: string;
  description?: string;
  price: number;
  tags: string[];
  createdAt: string;
}

// --- In-Memory Store (replace with real database) ---

const items: Map<string, Item> = new Map();

// Seed with demo data.
const demoId = "550e8400-e29b-41d4-a716-446655440000";
items.set(demoId, {
  id: demoId,
  name: "CopilotForge Starter Kit",
  description: "Everything you need to get started",
  price: 29.99,
  tags: ["starter", "toolkit"],
  createdAt: new Date().toISOString(),
});

// --- Route Handlers ---

const router = Router();

/** GET /items — List items with pagination and search. */
router.get("/items", validate(listQuerySchema, "query"), (req: Request, res: Response) => {
  const query = req.query as unknown as ListQuery;

  let results = Array.from(items.values());

  // Filter by search term.
  if (query.search) {
    const term = query.search.toLowerCase();
    results = results.filter(
      (item) =>
        item.name.toLowerCase().includes(term) ||
        item.description?.toLowerCase().includes(term)
    );
  }

  // Paginate.
  const total = results.length;
  const start = (query.page - 1) * query.pageSize;
  const paged = results.slice(start, start + query.pageSize);

  res.json({
    data: paged,
    pagination: {
      page: query.page,
      pageSize: query.pageSize,
      total,
      totalPages: Math.ceil(total / query.pageSize),
    },
  });
});

/** GET /items/:id — Get a single item by ID. */
router.get("/items/:id", validate(itemIdSchema, "params"), (req: Request, res: Response, next: NextFunction) => {
  const { id } = req.params;
  const item = items.get(id);

  if (!item) {
    return next(new RouteError(`Item '${id}' not found`, 404, "NOT_FOUND"));
  }

  res.json({ data: item });
});

/** POST /items — Create a new item. */
router.post("/items", validate(createItemSchema, "body"), (req: Request, res: Response) => {
  const input = req.body as CreateItemInput;

  // TODO: Replace with real ID generation (e.g., uuid).
  const id = crypto.randomUUID();

  const item: Item = {
    id,
    name: input.name,
    description: input.description,
    price: input.price,
    tags: input.tags,
    createdAt: new Date().toISOString(),
  };

  items.set(id, item);
  console.log(`[route] Created item: ${item.name} (${id})`);

  res.status(201).json({ data: item });
});

/** PATCH /items/:id — Update an existing item. */
router.patch(
  "/items/:id",
  validate(itemIdSchema, "params"),
  validate(updateItemSchema, "body"),
  (req: Request, res: Response, next: NextFunction) => {
    const { id } = req.params;
    const item = items.get(id);

    if (!item) {
      return next(new RouteError(`Item '${id}' not found`, 404, "NOT_FOUND"));
    }

    const input = req.body as UpdateItemInput;

    // Apply partial updates.
    if (input.name !== undefined) item.name = input.name;
    if (input.description !== undefined) item.description = input.description;
    if (input.price !== undefined) item.price = input.price;
    if (input.tags !== undefined) item.tags = input.tags;

    console.log(`[route] Updated item: ${item.name} (${id})`);
    res.json({ data: item });
  }
);

/** DELETE /items/:id — Delete an item. */
router.delete("/items/:id", validate(itemIdSchema, "params"), (req: Request, res: Response, next: NextFunction) => {
  const { id } = req.params;

  if (!items.has(id)) {
    return next(new RouteError(`Item '${id}' not found`, 404, "NOT_FOUND"));
  }

  items.delete(id);
  console.log(`[route] Deleted item: ${id}`);

  res.status(204).send();
});

// --- App Setup ---

async function main() {
  const app = express();
  app.use(express.json());

  // Mount routes.
  app.use("/api", router);

  // Error handler must be last.
  app.use(errorHandler);

  // TODO: Replace port with your preferred port.
  const PORT = parseInt(process.env.PORT ?? "3200", 10);
  app.listen(PORT, () => {
    console.log("=== Route Handler Recipe ===\n");
    console.log(`Server listening on http://localhost:${PORT}`);
    console.log("\nRoutes:");
    console.log("  GET    /api/items          — list with pagination & search");
    console.log("  GET    /api/items/:id      — get single item");
    console.log("  POST   /api/items          — create item");
    console.log("  PATCH  /api/items/:id      — update item");
    console.log("  DELETE /api/items/:id      — delete item");
    console.log(`\nTry: curl http://localhost:${PORT}/api/items`);
  });
}

main().catch(console.error);

File: cookbook/route-handler.py
"""
route-handler.py — CopilotForge Cookbook Recipe

WHAT THIS DOES:
    Demonstrates a FastAPI route handler with Pydantic input validation,
    dependency injection, proper HTTP status codes, and structured error
    responses. Shows patterns for GET (list/detail), POST (create),
    and PATCH (update) endpoints.

WHEN TO USE THIS:
    When building FastAPI applications and you want consistent request
    validation, error responses, and typed route handlers.

HOW TO RUN:
    1. pip install fastapi uvicorn
    2. uvicorn cookbook.route-handler:app --reload
    (Or: python cookbook/route-handler.py)

PREREQUISITES:
    - Python 3.10+
    - fastapi >= 0.100
    - uvicorn >= 0.23
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Annotated

import uvicorn
from fastapi import Depends, FastAPI, HTTPException, Query, status
from pydantic import BaseModel, Field

# --- Schemas ---


class CreateItemRequest(BaseModel):
    name: str = Field(min_length=1, max_length=100, description="Item name")
    description: str | None = Field(default=None, max_length=500, description="Optional description")
    price: float = Field(gt=0, description="Price must be positive")
    tags: list[str] = Field(default_factory=list, max_length=10, description="Up to 10 tags")


class UpdateItemRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    description: str | None = Field(default=None, max_length=500)
    price: float | None = Field(default=None, gt=0)
    tags: list[str] | None = Field(default=None, max_length=10)


class ItemResponse(BaseModel):
    id: str
    name: str
    description: str | None = None
    price: float
    tags: list[str]
    created_at: str


class PaginationInfo(BaseModel):
    page: int
    page_size: int
    total: int
    total_pages: int


class ItemListResponse(BaseModel):
    data: list[ItemResponse]
    pagination: PaginationInfo


class ErrorDetail(BaseModel):
    code: str
    message: str
    details: list[dict[str, str]] | None = None


class ErrorResponse(BaseModel):
    error: ErrorDetail


# --- In-Memory Store (replace with real database) ---

items_store: dict[str, ItemResponse] = {}

# Seed with demo data.
_demo_id = "550e8400-e29b-41d4-a716-446655440000"
items_store[_demo_id] = ItemResponse(
    id=_demo_id,
    name="CopilotForge Starter Kit",
    description="Everything you need to get started",
    price=29.99,
    tags=["starter", "toolkit"],
    created_at=datetime.now(timezone.utc).isoformat(),
)

# --- Dependencies ---


class ListParams:
    """Dependency for pagination and search query params."""

    def __init__(
        self,
        page: Annotated[int, Query(ge=1, description="Page number")] = 1,
        page_size: Annotated[int, Query(ge=1, le=100, alias="pageSize", description="Items per page")] = 20,
        search: Annotated[str | None, Query(description="Search by name or description")] = None,
    ) -> None:
        self.page = page
        self.page_size = page_size
        self.search = search


# --- Helper ---


def get_item_or_404(item_id: str) -> ItemResponse:
    """Look up an item by ID, raising 404 if not found."""
    item = items_store.get(item_id)
    if item is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": "NOT_FOUND", "message": f"Item '{item_id}' not found"},
        )
    return item


# --- App & Routes ---

app = FastAPI(
    title="Route Handler Example",
    responses={
        400: {"model": ErrorResponse, "description": "Validation error"},
        404: {"model": ErrorResponse, "description": "Not found"},
    },
)


@app.get("/api/items", response_model=ItemListResponse)
async def list_items(params: ListParams = Depends()) -> ItemListResponse:
    """List items with pagination and optional search."""
    results = list(items_store.values())

    # Filter by search term.
    if params.search:
        term = params.search.lower()
        results = [
            item
            for item in results
            if term in item.name.lower() or (item.description and term in item.description.lower())
        ]

    # Paginate.
    total = len(results)
    start = (params.page - 1) * params.page_size
    paged = results[start : start + params.page_size]

    return ItemListResponse(
        data=paged,
        pagination=PaginationInfo(
            page=params.page,
            page_size=params.page_size,
            total=total,
            total_pages=max(1, -(-total // params.page_size)),  # Ceiling division.
        ),
    )


@app.get("/api/items/{item_id}", response_model=dict[str, ItemResponse])
async def get_item(item_id: str) -> dict[str, ItemResponse]:
    """Get a single item by ID."""
    item = get_item_or_404(item_id)
    return {"data": item}


@app.post("/api/items", response_model=dict[str, ItemResponse], status_code=status.HTTP_201_CREATED)
async def create_item(body: CreateItemRequest) -> dict[str, ItemResponse]:
    """Create a new item."""
    # TODO: Replace with real ID generation and database insert.
    item_id = str(uuid.uuid4())

    item = ItemResponse(
        id=item_id,
        name=body.name,
        description=body.description,
        price=body.price,
        tags=body.tags,
        created_at=datetime.now(timezone.utc).isoformat(),
    )

    items_store[item_id] = item
    print(f"[route] Created item: {item.name} ({item_id})")

    return {"data": item}


@app.patch("/api/items/{item_id}", response_model=dict[str, ItemResponse])
async def update_item(item_id: str, body: UpdateItemRequest) -> dict[str, ItemResponse]:
    """Update an existing item (partial update)."""
    item = get_item_or_404(item_id)

    # Apply partial updates by creating a new response with merged fields.
    update_data = body.model_dump(exclude_unset=True)
    updated = item.model_copy(update=update_data)

    items_store[item_id] = updated
    print(f"[route] Updated item: {updated.name} ({item_id})")

    return {"data": updated}


@app.delete("/api/items/{item_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_item(item_id: str) -> None:
    """Delete an item."""
    get_item_or_404(item_id)  # Raises 404 if not found.
    del items_store[item_id]
    print(f"[route] Deleted item: {item_id}")


# --- Startup ---


def main() -> None:
    print("=== Route Handler Recipe ===\n")
    # TODO: Replace host/port with your preferred values.
    uvicorn.run(app, host="0.0.0.0", port=8100)


if __name__ == "__main__":
    main()
